### DISCUSSION TITLE
Question about scanning speed and what to do to speed it up.

### DISCUSSION THREAD

**Discussion by gahabana** [MoleMan1024/audiowagon#101] [General]
Hi, initially i put 40,000 songs on USB stick in Opus format. Worked fine but scanning took really long time and I could never (estimated) spend 8 hours on roadtrip in one go to complete it. 

In the meantime I put those 40k songs on USB/NVME drive which did occupy 1.1TB in FLAC format. Scanning seems little faster but after one hour in the car I have reached about 6000 songs. 34k to go. 

I have Volvo XC60 T8 late 2022 ... is there anything i can try to do (except lower size of the library to 1-5k songs) to make scanning of the library actually complete ?

As for speed of the drives - both were formatted as FAT32, DOS partition scheme, single primary partition, 500GB in 1st case and 2TB in the second with FAT32 filesystem on it. I checked cluster size and it was 32k in both cases ... there doesn't seem to be anything else I could do.

Is there any advice or speed of the CPU, AAOS, libraries used by AudioWagon create a limitation that can not be overcome ?

thanks in advance !!!

**Comment by MoleMan1024** [comment]
Hi,
I assume you have read [my thoughts on this](https://moleman1024.github.io/audiowagon/faq.html#is-there-a-limitation-on-the-number-of-files-the-app-can-handle) and also know there are [settings to turn off indexing for large libraries](https://moleman1024.github.io/audiowagon/faq.html#what-happens-during-indexing)? 

Assuming you would still like your whole 40k library fully indexed in your car... well...

- This is not so much CPU bound, but an I/O bound operation. You already tried NVME which is supposedly the fastest thing we have for USB. 
- If you have logging to USB turned on, please turn it off, that will improve the speed a bit. Also avoid playing from the USB drive while indexing would speed it up slightly.
- The USB access can't be parallelized easily due to limitations in the library. If someone were to put more effort into this it could maybe be sped up more.
- I don't recall what I have measured before but 8 hours for 40k songs seems too long, that would be 1.4 seconds per file, seems a bit high... maybe send me a log file so I can see how long each operation on each file takes based on timestamps. 
- I have never really measured with .flac files either I mostly try mp3s. .flac are bigger but the metadata should be in the headers so it should not be that much slower to read those...
- it could be indexed outside the car if you built the source code e.g. on a dedicated Android phone or change the code to make it work on a desktop PC but that is a bit involved

**Reply by gahabana** [reply]
hi, thanks again for prompt response - will check if I have indeed turned on USB logging. I don't think so but if so i'd be happy since that would easily explain the issue !
BTW, what i did observer and think this is good ... after an hour of driving with Opus files i had about 4-5 thousands indexed and then i stopped the car and when i turned it on it seemed to start from 0 ... however after few minutes it was already again at 4-5k which i think implies - whatever it indexed 1st time it did save it to sqliteDB file on USB drive (or whatever is being used) so upon 2nd turned on it was wee bit faster ... 

Will come back and thanks again !

**Reply by gahabana** [reply]
hi, i did more experimenting with 2 NVME-to-USB-C enclosures. I tried also formatting the drives under Windows/Rufus (rather then Linux) to make sure partitioning/formatting is not the culprit. 
I do not have logging to USB turned on ... and I have also 'deleted' the database(s) from the 'internal' drives within the car (as drives were cleaned/reformatted and songs copied).

Nevertheless after 5 minutes and having (either through 'index manually' or 'automatically' ) after 5 minutes and getting to 46 songs I called it quits. 

I am not sure what else to do ... maybe I should turn on USB logging , start manual scan and after e.g. 10 minutes disconnect the drive (turn off the car 1st) and post here log files ?

Thanks in advance !!!

**Reply by MoleMan1024** [reply]
yes please turn on USB logging, insert USB drive, let it index for a bit and then eject it from AudioWagon settings menu. Provide resulting log file, maybe I can see where all the time is going.

**Reply by gahabana** [reply]
hi, done as per above. 1st time i did exactly as suggested BUT simply switched of the car rather then 'ejecting' the drive and found no log files on a drive. 2nd time i put disk in, prior to that deleted all the DB's and turned USB logging on. it started indexing. Few mins later it was at 500 songs and i stopped USB logging then i ejected the drive and I could not find any written files on the drive.

Could it be that Audiowagon can not WRITE to a these drives and that also implies some other issues (hence issue with speed) ? I get no crashes, no error messages (if it is 'good news').... ?

tx

**Reply by MoleMan1024** [reply]
If no entries can be written to your disk it may indeed be a problem with the disks or the adapter you use. You can also view the logging by tapping on the "version" field in the AudioWagon settings, it is a little cumbersome but you could take photos.
Without log files I can't tell what is going on.

**Reply by gahabana** [reply]
will try that. in the meantime i saw double directories on the drive that was in the car (that is on macos) ... then i checked if there are any special limitations on FAT32 root directory and it seems (though neither windows nor linux complained) that root directory can not have more then 512 entries and all my artists (which are the 1st directory) were 900 long.
my guess is that is what has caused the issue - will check tomorrow ... i supposed i need to make top level directory 'music' and seems FAT32 has limitation that non-ROOT directories have limit of 2048 entries so issue should go away.

**Reply by MoleMan1024** [reply]
I don't know where you got this info from, FAT32 should support multiple thousand files per directory, see [here](https://superuser.com/a/1544848) . You can try it of course, but unless there is another bug in libaums (the library I use for the USB filesystem) this should not be a problem. Maybe you looked at [FAT16](http://www.qnx.com/developers/docs/qnxcar2/index.jsp?topic=%2Fcom.qnx.doc.neutrino.user_guide%2Ftopic%2Flimits_FAT.html) ?

**Reply by gahabana** [reply]
i found few sources stating that ... .however, i assume this one from 2000 is 'the reference' doc as it comes from Microsoft (link at the bottom of the page):
https://answers.microsoft.com/en-us/windows/forum/all/what-is-the-maximum-number-of-files-i-can-place/9b371fae-5e09-4fe6-b24f-be16547896f3#:~:text=The%20maximum%20size%20for%20a,be%20avoided%20for%20performance%20reasons.
it does confirm that you are 100% correct:

![image](https://user-images.githubusercontent.com/881265/211526791-9ac65fc8-eafd-49cf-9df6-5ce9224d2bf5.png)

I'll do more digging. Am trying to get larger USB key (512GB) and go with compressed/Opus library will report back once i have something usable. Thank you!

**Reply by gahabana** [reply]
> If no entries can be written to your disk it may indeed be a problem with the disks or the adapter you use. You can also view the logging by tapping on the "version" field in the AudioWagon settings, it is a little cumbersome but you could take photos. Without log files I can't tell what is going on.

Hi, i succeeded - reformatted the drive, made sure clusters were 32K on FAT32 etc ... did manage to get logs written to the log file.
Scanning speed was as slow as usual. I did not see any error messages. 
Hope you can tell from log file what might be the root cause. 

Log should address time from plugging in the drive for the 1st time, letting it scan for almost 15min ....
Here it is and thank you in advance !!!
[audiowagon_2023-01-10_14-56-36.log](https://github.com/MoleMan1024/audiowagon/files/10383719/audiowagon_2023-01-10_14-56-36.log)

**Reply by MoleMan1024** [reply]
Thanks for the log. It looks like the extraction of metadata and album art takes quite long for these .flac files, each one takes like 300 to 500 ms. With 40k songs indexing at this speed it would take around 5,6 hours.
I will see if I can get some .flac files and see if I can reproduce this slowness with my USB drives.

**Reply by gahabana** [reply]
I tried with Opus files (though i think they copied how FLAC stores tags and that scheme is supposed to be quite efficient) and speed is exactly the same. 
The only ones i have not tried with are MP3 and AAC format but with large drives I was hoping i can fit all my music in lossless format (FLAC) rather then compress them and Opus is supposedly about best sounding lossy codec.

If you need access to bunch of flac files i can help you access them remotely if you wish - let me know. But simplest for you might be to take whatever you have (e.g. MP3s ) and convert them to FLACs including metadata. I use foobar2000 but other tools will do this job as well.

tx again !

**Reply by gahabana** [reply]
short update - i've tried by removing all album cover tags from entire library. both Opus as well as FLAC. Scanning speed has remained at the level before 1-2 songs per second. 

Next step is to try mix of MP3 and M4A (AAC aka Apple music) files ... have about 1000 MP3s and 2000 M4As ... 22GB which should fit on small 32GB stick. Will check and report back

**Reply by gahabana** [reply]
hi, now i've tried all options. 
With AAC files (.m4a extension) reading of tags is on par or little faster than with FLAC and Opus files (60 seconds for 80 files) .
The ONLY one where it works reasonably fast is MP3 files. in 12 seconds, 100 files, 60 seconds, ~500 files on Volvo 2022. 
So even for library of 40,000 files it would take 5000 seconds which is 80 minutes which is kinda bearable ...

So my conclusion is - unless there is something 'odd' in how AudioWagon is processing tags culrpit must be underlying metadata "MediaMetadataRetriever" library which does good job with MP3s but sucks with all other supported media formats.

**Reply by gahabana** [reply]
just a thought - before changes you are making - scanning speed is about (for non MP3s) is about 0.6sec/file and on MP3s is more like 0.14sec/file (or 1.5 files per second and 7 files per sec respectively).... so MP3 is 5 times faster.
MP3 tags (in my case) seem to be at the beginning (ID3v2.3 tags) of a file where as for other formats it is also at the beginning of the file to help 'streaming' apps get all the tags before audio content is streamed. I've checked both FLAc and Opus. So this does not explain why it is so much faster with MP3s.

I've googled a lot and used ChatGPT to try to find some answers and there are 2 suggestions that you may consider. I am sceptical of 1st since it may mean entire file is being read but for some folks that seems to improve speed and is easy to test for you. Latter is bit more involved but seems not to hard (apart from potentially breaking something else) and am not sure you'd want to try that:

Option 1:
    Use the setDataSource(FileDescriptor fd, long offset, long length) method instead of setDataSource(String path) method. This will allow you to pass a file descriptor of the file and the offset and length of the data you want to extract metadata from.
for example: 
```
FileInputStream inputStream = new FileInputStream(file);
FileDescriptor fd = inputStream.getFD();
MediaMetadataRetriever mmr = new MediaMetadataRetriever();
mmr.setDataSource(fd, 0, file.length());
```
Option 2:
Use MediaExtractor class which is lower level (needs bit more work). E.g.:
```

MediaExtractor extractor = new MediaExtractor();
extractor.setDataSource(filePath);

int numTracks = extractor.getTrackCount();
for (int i = 0; i < numTracks; i++) {
    MediaFormat format = extractor.getTrackFormat(i);
    String mime = format.getString(MediaFormat.KEY_MIME);
    if (mime.startsWith("audio/")) {
        // Extract metadata
        String title = format.getString(MediaFormat.KEY_TITLE);
        String artist = format.getString(MediaFormat.KEY_ARTIST);
        String album = format.getString(MediaFormat.KEY_ALBUM);
        // Do something with the metadata
    }
}
extractor.release();

```

**Reply by MoleMan1024** [reply]
Option 1) is essentially what I am doing however you cannot use the Java file API with libaums so the code you posted will not work as it is. Reading the whole file when you just want the metadata is wasteful and will be slower. 

Option 2) does not make sense to me as the MediaFormat keys you mention here [do not exist](https://developer.android.com/reference/kotlin/android/media/MediaFormat). 

Have some patience, I am working on some optimizations.

**Reply by gahabana** [reply]
> Option 1) is essentially what I am doing however you cannot use the Java file API with libaums so the code you posted will not work as it is. Reading the whole file when you just want the metadata is wasteful and will be slower.
> 
> Option 2) does not make sense to me as the MediaFormat keys you mention here [do not exist](https://developer.android.com/reference/kotlin/android/media/MediaFormat).
> 
> Have some patience, I am working on some optimizations.

You're (again/always) right :)
 ... will stop with useless recommendations and hope you find a way  :) !!!

**Reply by petard** [reply]
For reference, it took about 15min to index an SSD with 19.000. Now I was wondering what do with incremental updates, for example after adding some new music.

I was thinking if there is a way to index and create the database on my computer so AudioWagon would only have to read the database file. Or do I need to index each new folder manually if I don't want to scan the whole disk again?

**Reply by MoleMan1024** [reply]
@petard : Please read [this section from the FAQ about how indexing works](https://moleman1024.github.io/audiowagon/faq.html#why-does-it-take-so-long-when-starting-the-app) and this [suggestion](https://moleman1024.github.io/audiowagon/faq.html#is-there-a-limitation-on-the-number-of-files-the-app-can-handle) . I am not planning to try to optimize the speed any more right now.
